types/transaction.go

```diff
@@ -12,73 +13,102 @@ import (
 type TxType byte
 
 const (
-	LegacyTx TxType = 0x0
-	StateTx  TxType = 0x7f
+	LegacyTxType TxType = 0x0
+	StateTxType  TxType = 0x7f
+	DynamicFeeTx TxType = 0x8f
 
 	StateTransactionGasLimit = 1000000 // some arbitrary default gas limit for state transactions
 )
 
 func ReadRlpTxType(rlpValue *fastrlp.Value) (TxType, error) {
 	bytes, err := rlpValue.Bytes()
 	if err != nil {
-		return LegacyTx, err
+		return LegacyTxType, err
 	}
 
 	if len(bytes) != 1 {
-		return LegacyTx, fmt.Errorf("expected 1 byte transaction type, but size is %d", len(bytes))
+		return LegacyTxType, fmt.Errorf("expected 1 byte transaction type, but size is %d", len(bytes))
 	}
 
 	b := TxType(bytes[0])
 
 	switch b {
-	case LegacyTx, StateTx:
+	case LegacyTxType, StateTxType:
 		return b, nil
 	default:
-		return LegacyTx, fmt.Errorf("invalid tx type value: %d", bytes[0])
+		return LegacyTxType, fmt.Errorf("invalid tx type value: %d", bytes[0])
 	}
 }
 
 func (t TxType) String() (s string) {
 	switch t {
-	case LegacyTx:
+	case LegacyTxType:
 		return "LegacyTx"
-	case StateTx:
+	case StateTxType:
 		return "StateTx"
 	default:
 		return "UnknownTX"
 	}
 }
 
+type TxData interface {
+	txType() TxType // returns the type ID
+	Copy() TxData   // creates a deep copy and initializes all fields
+
+	input() []byte
+	gas() uint64
+	setGas(gas uint64)
+	gasPrice() *big.Int
+	gasTipCap() *big.Int
+	gasFeeCap() *big.Int
+	value() *big.Int
+	nonce() uint64
+	to() *Address
+
+	rawSignatureValues() (v, r, s *big.Int)
+	setSignatureValues(v, r, s *big.Int)
+}
+
 type Transaction struct {
-	Nonce    uint64
-	GasPrice *big.Int
-	Gas      uint64
-	To       *Address
-	Value    *big.Int
-	Input    []byte
-	V        *big.Int
-	R        *big.Int
-	S        *big.Int
-	Hash     Hash
-	From     Address
-
-	Type TxType
-
-	// Cache
+	inner TxData    // Consensus contents of a transaction
+	time  time.Time // Time first seen locally (spam avoidance)
+
+	// Caches
+	hash Hash
 	size atomic.Value
+	from Address
+}
+
+// NewTx creates a new transaction.
+func NewTx(inner TxData) *Transaction {
```

Comment:
when we start using go 1.19 it'd be nice to use generics here and in NewTxWithSender
Please, make a task for it.